src/index/IndexImpl.cpp

```diff
@@ -0,0 +1,1772 @@
+      return fLifted(OSP());
+    case Permutation::OPS:
+      return fLifted(OPS());
+    default:
+      AD_CHECK(false);
+  }
+}
+
+// _____________________________________________________________________________
+template <class PermutationImplementation>
+vector<float> IndexImpl::getMultiplicitiesImpl(
+    const PermutationImplementation& p, const string& key) const {
+  Id keyId;
+  vector<float> res;
+  if (_vocab.getId(key, &keyId) && p._meta.relationExists(keyId)) {
+    auto rmd = p._meta.getRmd(keyId);
+    auto logM1 = rmd.getCol1LogMultiplicity();
+    res.push_back(static_cast<float>(pow(2, logM1)));
+    auto logM2 = rmd.getCol2LogMultiplicity();
+    res.push_back(static_cast<float>(pow(2, logM2)));
+  } else {
+    res.push_back(1);
+    res.push_back(1);
+  }
+  return res;
+}
+
+// _____________________________________________________________________________________________
+vector<float> IndexImpl::getMultiplicities(const string& key,
+                                           const Permutation p) const {
+  auto f = [&](const auto& permutationImpl) {
+    return getMultiplicitiesImpl(permutationImpl, key);
+  };
+  return applyForPermutation(p, f);
+}
+
+// ___________________________________________________________________
+template <class PermutationImplementation>
+vector<float> IndexImpl::getMultiplicitiesImpl(
+    const PermutationImplementation& p) const {
+  std::array<float, 3> m{
+      static_cast<float>(getNofTriples() / getNofSubjects()),
+      static_cast<float>(getNofTriples() / getNofPredicates()),
+      static_cast<float>(getNofTriples() / getNofObjects())};
+
+  return {m[p._keyOrder[0]], m[p._keyOrder[1]], m[p._keyOrder[2]]};
+}
+
+// _____________________________________________________________________________
+vector<float> IndexImpl::getMultiplicities(const Permutation p) const {
+  auto f = [this](const auto& permutationImpl) {
+    return getMultiplicitiesImpl(permutationImpl);
+  };
+  return applyForPermutation(p, f);
+}
+
+// ___________________________________________________________________________
+template <PermutationConcept P>
+void IndexImpl::scanImpl(Id key, IdTable* result, const P& p,
+                         ad_utility::SharedConcurrentTimeoutTimer timer) const {
+  if (p._meta.relationExists(key)) {
+    const FullRelationMetaData& rmd = p._meta.getRmd(key)._rmdPairs;
+    result->reserve(rmd.getNofElements() + 2);
+    result->resize(rmd.getNofElements());
+    p._file.read(result->data(), rmd.getNofElements() * 2 * sizeof(Id),
+                 rmd._startFullIndex, std::move(timer));
+  }
+}
+
+// ___________________________________________________________________________
+void IndexImpl::scan(Id key, IdTable* result, const Permutation& p,
+                     ad_utility::SharedConcurrentTimeoutTimer timer) const {
+  auto f = [&](const auto& permutationImpl) {
+    return scanImpl(key, result, permutationImpl, timer);
+  };
+  return applyForPermutation(p, f);
+}
+
```

Comment:
The `scanImpl` is the previous `scan` method,
And the new `scan` method just does the Transition from the simple `Permutation` enum,
to the different `PermutationImpl` types

(similar for the two other overloads below)